src/command.ml

```diff
@@ -0,0 +1,118 @@
+open! Stdune
+open Import
+
+type static = Static
+type dynamic = Dynamic
+
+type _ t =
+  | A        : string -> _ t
+  | As       : string list -> _ t
+  | S        : 'a t list -> 'a t
+  | Concat   : string * 'a t list  -> 'a t
+  | Dep      : Path.t -> _ t
+  | Deps     : Path.t list -> _ t
+  | Target   : Path.t -> dynamic t
+  | Path     : Path.t -> _ t
+  | Paths    : Path.t list -> _ t
+  | Hidden_deps    : Dep.Set.t -> _ t
+  | Hidden_targets : Path.t list -> dynamic t
+  | Dyn      : static t Build.s -> dynamic t
+  | Fail     : fail -> _ t
+
+let dyn_args args = Dyn (Build.S.map args ~f:(fun x -> As x))
+
+let dyn_deps args = Dyn (Build.S.map args ~f:(fun x -> Deps x))
+let dyn_dep  args = Dyn (Build.S.map args ~f:(fun x -> Dep x))
+
+let rec add_targets ts acc =
+  List.fold_left ts ~init:acc ~f:(fun acc t ->
+    match t with
+    | Target fn  -> fn :: acc
+    | Hidden_targets fns -> List.rev_append fns acc
+    | S ts
+    | Concat (_, ts) -> add_targets ts acc
+    | _ -> acc)
+
+let expand ~dir ts =
+  let run_loop t =
+    let static_deps = ref Dep.Set.empty in
+    let add_dep path = static_deps := Dep.Set.add !static_deps (Dep.file path) in
+    let rec loop_static : static t -> string list  = function
+      | A s  -> [s]
+      | As l -> l
+      | Dep fn ->
+        add_dep fn;
+        [Path.reach fn ~from:dir]
+      | Path fn -> [Path.reach fn ~from:dir]
+      | Deps fns ->
+        List.map fns ~f:(fun fn ->
+          add_dep fn;
+          Path.reach ~from:dir fn)
+      | Paths fns ->
+        List.map fns ~f:(Path.reach ~from:dir)
+      | S ts -> List.concat_map ts ~f:loop_static
+      | Concat (sep, ts) -> [String.concat ~sep (loop_static (S ts))]
+      | Hidden_deps l ->
+        static_deps := Dep.Set.union !static_deps l;
+        []
+      | Fail f -> f.fail ()
+    in
+    (loop_static t, !static_deps)
```

Comment:
Argh, it took ages for me to find this. This dereferences `!static_deps` and runs the loop in unspecified order.
You want to write:
```
let res = loop_static t in
(res, !static_deps)
```

Reply:
Ouch! Thanks for spotting -- fixing in the next commit.